From 1556bb275f87f7860511b8cd81a53c77265c8fad Mon Sep 17 00:00:00 2001
From: Raine Makelainen <raine.makelainen@jolla.com>
Date: Tue, 28 Nov 2017 15:32:52 +0200
Subject: [PATCH 1/4] Fix acinclude mktemp

Signed-off-by: Raine Makelainen <raine.makelainen@jolla.com>
---
 acinclude.m4 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/acinclude.m4 b/acinclude.m4
index 5c54d4a..df638f3 100644
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -65,7 +65,7 @@ dnl file that just includes the HEADER using the given CFLAGS. In case of
 dnl failure, the ERR_MSG will be printed using the LIBBLOCKDEV_FAILURES macro.
 AC_DEFUN([LIBBLOCKDEV_CHECK_HEADER], [dnl
 echo -n "Checking header [$1] existence and usability..."
-temp_file=$(mktemp --tmpdir XXXXX.c)
+temp_file=$(mktemp)
 echo "#include <$1>" > $temp_file
 ${CC} -c [$2] $temp_file
 status=$?
-- 
2.7.4

