From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Vojtech Trefny <vtrefny@redhat.com>
Date: Fri, 12 Jul 2019 13:07:27 +0200
Subject: [PATCH] Redirect libfdisk log messages to our log

---
 src/plugins/part.c | 30 ++++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/src/plugins/part.c b/src/plugins/part.c
index 0056905..fd2fd69 100644
--- a/src/plugins/part.c
+++ b/src/plugins/part.c
@@ -182,6 +182,8 @@ static const PartFlag part_flags[18] = {
 #define _PART_NTFS        0x07
 #define _PART_FLAG_HIDDEN 0x10
 
+#define UNUSED __attribute__((unused))
+
 /**
  * get_part_num: (skip)
  *
@@ -215,6 +217,33 @@ static gint get_part_num (const gchar *part, GError **error) {
     return part_num;
 }
 
+static int fdisk_ask_callback (struct fdisk_context *cxt UNUSED, struct fdisk_ask *ask, void *data UNUSED) {
+    gint type = 0;
+    const gchar *fdisk_msg = NULL;
+    gchar *message = NULL;
+
+    type = fdisk_ask_get_type (ask);
+    fdisk_msg = fdisk_ask_print_get_mesg (ask);
+
+    switch (type) {
+        case FDISK_ASKTYPE_INFO:
+            message = g_strdup_printf ("[fdisk] %s", fdisk_msg);
+            bd_utils_log (LOG_INFO, message);
+            g_free (message);
+            break;
+        case FDISK_ASKTYPE_WARNX:
+        case FDISK_ASKTYPE_WARN:
+            message = g_strdup_printf ("[fdisk] %s", fdisk_msg);
+            bd_utils_log (LOG_WARNING, message);
+            g_free (message);
+            break;
+        default:
+            break;
+    }
+
+    return 0;
+}
+
 static struct fdisk_context* get_device_context (const gchar *disk, GError **error) {
     struct fdisk_context *cxt = fdisk_new_context ();
     gint ret = 0;
@@ -234,6 +263,7 @@ static struct fdisk_context* get_device_context (const gchar *disk, GError **err
     }
 
     fdisk_disable_dialogs(cxt, 1);
+    fdisk_set_ask (cxt, fdisk_ask_callback, NULL);
     return cxt;
 }
 
-- 
2.33.1

